samples: net: irc_bot: simplify and update irc_bot sample

mike-scott opened the pull request:
Originally, irc_bot sample was intended to be a prototype for setting
up an IRC-based parsing / handler library in subsys/net/lib.  For this
purpose it has several structures (irc and channel related) to allow
for individual channel based message handlers.

This effort has faded over time and the sample itself ended up being
very complex for the average user to "play around" with.

Let's simplify this sample by removing the irc/channel structures
and update it to use the full net_app APIs.  Also, fix the parse
issues across multiple net bufs.

These changes remove over 400 lines of code from this sample (~30%
size reduction).

Signed-off-by: Michael Scott <michael.scott@linaro.org>

mike-scott commented:
I considered breaking this up into several small patches, but they all seemed to touch each other so I ended up with 1 rather large patch.  :/  Recommendations are welcome.

lpereira commented on samples/net/irc_bot/src/irc-bot.c:
```diff
@@ -43,179 +45,144 @@ static u8_t cmd_buf[CMD_BUFFER_SIZE];
-#define DEFAULT_SERVER	"irc.freenode.net"
 #define DEFAULT_PORT	6667
 #define DEFAULT_CHANNEL	"#zephyrbot"
+#define MAX_CHANNELS	1
```
Unused `#define`.

mike-scott replied:
Heh, left over from cleanup.  Good catch.

lpereira commented on samples/net/irc_bot/src/irc-bot.c:
```diff
@@ -43,179 +45,144 @@ static u8_t cmd_buf[CMD_BUFFER_SIZE];
 
-	/* TODO: could be a private message (from another user) */
-	NET_INFO("Unknown privmsg received: %.*s\n", (int)len, cmd);
+	return;
```
Stray `return;`?

mike-scott replied:
Not sure how that got in there.. probably the same.  Thanks!

lpereira commented on samples/net/irc_bot/src/irc-bot.c:
```diff
@@ -240,409 +207,233 @@ process_command(struct zirc *irc, char *cmd, size_t len)
 			cmd_buf[cmd_len] = '\0';
-			process_command(irc, (char *)cmd_buf, cmd_len);
+			process_command((char *)cmd_buf, cmd_len);
+			/* skip the /n char after /r */
```
Shouldn't this be `\n after \r`?

mike-scott replied:
Fixing.

lpereira commented on samples/net/irc_bot/src/irc-bot.c:
```diff
@@ -834,239 +632,39 @@ on_msg_rcvd(void *data, struct zirc_chan *chan, char *umask, char *msg)
-
-	NET_INFO("Zephyr IRC bot sample");
+	memset(&app_ctx, 0, sizeof(app_ctx));
+	memset(cmd_buf, 0, CMD_BUFFER_SIZE);
```
These two are zeroed out as part of clearing the BSS, no need to memset() them here.

mike-scott replied:
Ugh, I had these as parameters and moved them out as statics later.  Another good catch.

lpereira reviewed (changes requested):
Agreed -- this is a good improvement. Some nits, though.

Commit:
samples: net: irc-bot: migrate to net_app API

Let's remove all of the extra coding to handle DHCP, DNS and event
management.  This is all handled by the net_app API.

Signed-off-by: Michael Scott <michael.scott@linaro.org>

Commit:
samples: net: irc-bot: migrate from NET_LOG to SYS_LOG

Let's migrate from using NET_LOG_* functions to the SYS_LOG
facility which allows for granular control over the display
of info, debug and error messages unique to the file that
you're in.

Signed-off-by: Michael Scott <michael.scott@linaro.org>

Commit:
samples: net: irc-bot: add support for CONFIG_NET_CONTEXT_NET_PKT_POOL

Some L2 layers such as bluetooth need to have extra buffer pools to
hold copies of packets for use with TCP.  Let's add support for that
so that future enablement of other boards won't get bitten.

Signed-off-by: Michael Scott <michael.scott@linaro.org>

Commit:
samples: net: irc-bot: clean up prj.conf

Several settings were out-dated or now enabled elsewhere.

Signed-off-by: Michael Scott <michael.scott@linaro.org>

Commit:
samples: net: irc-bot: fix cmd parsing across multiple buffers

Let's simplify the parsing function which scans incoming traffic
for carriage returns and let the net_pkt_read() function cross
multiple buffers if needed.  Also, gracefully skip lines which
are longer than the command buffer.

Signed-off-by: Michael Scott <michael.scott@linaro.org>

Commit:
samples: net: irc-bot: declare banner text at the top

For style points let's declare the text shown at the beginning
of main() function to the top of our source as APP_BANNER so that
it's obvious and presents nicely.

Signed-off-by: Michael Scott <michael.scott@linaro.org>

Commit:
samples: net: irc-bot: move initialite_hardware() code to main()

The code is easier to follow this way.

Signed-off-by: Michael Scott <michael.scott@linaro.org>

Commit:
samples: net: irc-bot: remove use of zirc and zirc_chan

These structures were originally intended to prototype an IRC library
which would be later moved into the subsys/net/lib/irc folder.

Unfortunately, that effort has faded and they make this code much more
complex than necessary.  Let's remove them in favor of simplicity.

Signed-off-by: Michael Scott <michael.scott@linaro.org>

Commit:
samples: net: irc-bot: establish a static global for nickname

Let's create a static global var for nickname so that we can use it
later without having to pass it all the way through callbacks, etc.

Also limit the size to 16 bytes as IRC servers will truncate anything
beyond.

Signed-off-by: Michael Scott <michael.scott@linaro.org>

Commit:
samples: net: irc-bot: handle private messages in some commands

Users can private message the bot and the bot should respond directly
back to that user rather than rudely ignoring them.

Signed-off-by: Michael Scott <michael.scott@linaro.org>

Commit:
samples: net: irc-bot: fix help text buffer size

At 32 bytes the snprintk call to fill the help buffer was most often
running out of room and returning an error.  Let's expand it to 64.

Signed-off-by: Michael Scott <michael.scott@linaro.org>

mike-scott commented:
I took some more time and split this up for easy review.   @lpereira sorry about wasting some of your time during the first round.  I addressed all of your comments.

jukkar reviewed (approved):
Looks good, thanks Mike for fixing this!

mike-scott commented:
@lpereira let me know if you need any other changes.
